include/utils/string.hpp

```diff
@@ -77,6 +82,10 @@ namespace string_util {
   string strip(const string& haystack, char needle);
   string strip_trailing_newline(const string& haystack);
 
+  string ltrim(string value, function<bool(char)> pred);
```

Comment:
Could you elaborate?

Reply:
To be able to store a function/functor… `std::function` is using type erasure. However to do so, there is at least one virtual call and that's the first overhead. Moreover if you use a template, the compiler may inline the predicate and so it is more efficient.

This is just a small optimization but I learnt that if you don't need to store something that can be either a function or a functor, you should avoid using `std::function`.

Reply:
So you mean something like this?
```cpp
template <bool (*P)(char)>
```

Reply:
Nop, more something like this:
```cpp
/***
\tparam Pred a functor that take a char as parameter and should return true if the char should be removed.
***/
template <typename Pred>
string ltrim(string value, Pred&& pred);
```
and you call the predicate like a normal function.
```cpp
pred(character)
```

We can also use your solution but if we are doing this we can't use lambda capture if needed.

Reply:
So I didn't manage to implement this because we overload the trim functions:

```cpp
  string ltrim(string&& value, const char& needle = ' ');
  string rtrim(string&& value, const char& needle = ' ');
  string trim(string&& value, const char& needle = ' ');
```

If you know how to fix this, feel free to push a commit to this branch. Otherwise we'll just leave it like this.

Reply:
Done. It was a little trickier because of the overload but that should be okay now.

Reply:
Oh wow. Isn't that a little overkill? I feel like the minimal efficiency increase doesn't justify added complexity.

Reply:
It just substitution failure, it disables the function if the parameter is a char.
If you don't like that no problem just remove my commit.

Reply:
I know substitution failure is quite hard to understand but until c++20 and concepts, there is nothing better :/

Reply:
I'll remove the commit again. The syntax is actually quite intuitive to read but it definitely adds a lot of complexity to maintain.